### PR TITLE
NAVAND-979: represent imperial distances in miles down to 0.1 miles in miles

### PR DESCRIPTION
**Pull request by dzinad**
I'm not really sure what we've decided to do with British yards but this PR fixes the part with too large distances being displayed in feet.

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt [hunk 1]
```diff
@@ -16,6 +16,8 @@ import kotlin.math.roundToInt
  */
 object MapboxDistanceUtil {
 
+    private const val smallDistanceLowerBoundInMiles = 0.1
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:19]: constants must be `UPPER_CASE`

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt [hunk 2]
```diff
@@ -37,31 +39,35 @@ object MapboxDistanceUtil {
...
+            FormattingRange.SMALL -> formatDistanceAndSuffixForSmallUnit(
+                distanceInMeters,
+                roundingIncrement,
+                unitType,
+                context,
+                locale
+            )
+            FormattingRange.MEDIUM -> formatDistanceAndSuffixForLargeUnit(
+                distanceInMeters,
+                1,
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:59]: out of scope: do you know why does `roundingIncrement` is ignoring for Medium and Large values?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt [hunk 3]
```diff
@@ -234,4 +240,38 @@ object MapboxDistanceUtil {
...
     }
+
+    private enum class FormattingRange { INVALID, SMALL, MEDIUM, LARGE }
+
+    private fun getFormattingRange(
+        distanceInMeters: Double,
+        unitType: UnitType
+    ): FormattingRange {
+        if (distanceInMeters !in 0.0..Double.MAX_VALUE) {
+            return FormattingRange.INVALID
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:251]: could you also add tests for the `INVALID` format (`distance` is negative)

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,6 +25,7 @@ Mapbox welcomes participation and contributions from everyone.
   ```
 - Added guarantees that route progress with `RouteProgress#currentState == OFF_ROUTE` arrives earlier than `NavigationRerouteController#reroute` is called. [#6764](https://github.com/mapbox/mapbox-navigation-android/pull/6764)
 - Fixed a rare `java.lang.NullPointerException: Attempt to read from field 'SpeechAnnouncement PlayCallback.announcement' on a null object reference` crash in `PlayCallback.getAnnouncement`. [#6760](https://github.com/mapbox/mapbox-navigation-android/pull/6760)
+- Changed distance formatting: now all the imperial distances down to 0.1 miles will be represented in miles. [#6775](https://github.com/mapbox/mapbox-navigation-android/pull/6775)
```
Comment (RingerJK) [inline CHANGELOG.md:28]:
```suggestion
- Changed distance formatting: now all the imperial distances down to `0.1` miles will be represented in miles, otherwise feet. [#6775](https://github.com/mapbox/mapbox-navigation-android/pull/6775)
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt (hunk 2 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:59]: I can only guess. I think it only makes sense for small units since their values are large. For example, we don't care if it's 451 or 453 meters, while 11 vs 12 km is quite a difference. If the user passes `roundingIncrement = 5` meaning that they want to display 451 meter as 450 meters, it probably doesn't mean that they also want to display 5 km instead of 4 km. That's just my guess but I'd say the docs are not clear enough (we should specify that it's only valid for small distances).

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt (hunk 1 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:19]: Why on Earth doesn't ktlint check that? :)

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt (hunk 3 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:251]: Sure, always happy to add tests. :)

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt [hunk 5]
```diff
@@ -37,31 +39,35 @@ object MapboxDistanceUtil {
...
-            in smallDistanceUpperThresholdInMeters..mediumDistanceUpperThresholdInMeters -> {
-                formatDistanceAndSuffixForLargeUnit(distanceInMeters, 1, unitType, context, locale)
-            }
-            else -> {
-                formatDistanceAndSuffixForLargeUnit(distanceInMeters, 0, unitType, context, locale)
-            }
+        return when (getFormattingRange(distanceInMeters, unitType)) {
+            FormattingRange.INVALID -> formatDistanceAndSuffixForSmallUnit(
+                distanceInMeters,
+                roundingIncrement,
```
Comment (dzinad) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:45]: It was 0 before but in that case roundingIncrement was applied to an artificial zero distance (when the original one was negative). So when invoking 
```
formatDistance(
        distanceInMeters = -234.4,
        roundingIncrement = RoundingIncrement.INCREMENT_FIFTY,
        unitType = UnitType.METRIC,
        context
    )
```
we used to get 50 meters, not 0 meters. I changed that: now it's 0 meters.

**Comment by abhishek1508** [comment]
@dzinad The idea here is to replicate what [iOS](https://github.com/mapbox/mapbox-navigation-ios/blob/96d0fc0744d64d2a092f571e1583eeb86831a774/Sources/MapboxCoreNavigation/DistanceFormatter.swift#L78-L111) is doing. On Android, our default `RoundingIncrement` is always 50, whereas iOS maintains a table. 

cc @LukasPaczos

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt (hunk 2 above)
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:59]: yeah, that makes sense. The API seems a bit confusing:  it provides the opportunity to set this value, but later one is ignored. 😞

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt (hunk 1 above)
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt]: Unfortunately, this rule does not exist https://pinterest.github.io/ktlint/rules/standard/ 😄 adding and testing a custom `rule` for this will take time. it's simply not worth it

Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt]: I think detekt knows about this rule. Because I've definitely senn this check somewhere and I think it was detekt.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt (hunk 5 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt:45]: OK, I reverted it because there was a test for this behaviour.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt (hunk 1 above)
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/formatter/MapboxDistanceUtil.kt]: It was disabled some time ago at https://github.com/detekt/detekt/pull/3151 because of `Jetpack Compose` [contract](https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-api-guidelines.md#singletons-constants-sealed-class-and-enum-class-values).

**Comment by dzinad** [comment]
> @dzinad The idea here is to replicate what [iOS](https://github.com/mapbox/mapbox-navigation-ios/blob/96d0fc0744d64d2a092f571e1583eeb86831a774/Sources/MapboxCoreNavigation/DistanceFormatter.swift#L78-L111) is doing. On Android, our default RoundingIncrement is always 50, whereas iOS maintains a table. When trying to achieve parity, in my understanding it would automatically apply a patch for British yards as well.

Discussed internally that other aspects of the ticket will be addressed separately.
